### PR TITLE
TRUNK-5963: Order Service does not allow saving Order Groups with an OrderContext

### PR DESCRIPTION
**Pull request by IamMujuziMoses**
## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-5963

## Checklist: I completed these to help reviewers :)

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command
- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by IamMujuziMoses** [comment]
@mseaton  @dkayiwa could you please review this PR

File: api/src/test/java/org/openmrs/api/OrderServiceTest.java [hunk 1]
```diff
@@ -1036,8 +1036,8 @@ public void saveOrder_shouldNotAllowEditingAnExistingOrder() {
 	 */
 	@Test
 	public void getCareSettingByUuid_shouldReturnTheCareSettingWithTheSpecifiedUuid() {
```
Comment (mseaton) [inline api/src/test/java/org/openmrs/api/OrderServiceTest.java:1038]: Why has this test been changed?  It doesn't seem related to the changes being made, and unclear why this is needed.

File: api/src/test/java/org/openmrs/api/OrderServiceTest.java [hunk 2]
```diff
@@ -3961,6 +3961,48 @@ public void saveOrder_shouldAllowARetrospectiveOrderToCloseAnOrderThatExpiredInT
...
+		Order secondOrder = new OrderBuilder().withAction(Order.Action.NEW).withPatient(7).withConcept(10).withOrderer(1)
+			.withEncounter(3).withDateActivated(new Date()).withOrderType(17)
+			.withUrgency(Order.Urgency.ON_SCHEDULED_DATE).withScheduledDate(new Date()).build();
+
+		orderGroup.addOrder(firstOrder);
+		orderGroup.addOrder(secondOrder);
+
+		OrderType orderType = new OrderType();
+		orderGroup.setId(1);
+		orderType.setOrderTypeId(17);
```
Comment (mseaton) [inline api/src/test/java/org/openmrs/api/OrderServiceTest.java:3991]: There are over a dozen order types in the standard test dataset.  Just get an existing one here and use it.

**Review by mseaton — CHANGES_REQUESTED** [review]

File: api/src/test/java/org/openmrs/api/OrderServiceTest.java [hunk 3]
```diff
@@ -3961,6 +3961,48 @@ public void saveOrder_shouldAllowARetrospectiveOrderToCloseAnOrderThatExpiredInT
...
+
+		CareSetting careSetting = new CareSetting();
+		careSetting.setId(1);
+		careSetting.setName("Neonatal Clinic");
```
[the remark this answers is not in the document]
Comment (IamMujuziMoses) [reply api/src/test/java/org/openmrs/api/OrderServiceTest.java:3995]: okay, let me fix that

File: api/src/main/java/org/openmrs/api/OrderService.java [hunk 4]
```diff
@@ -834,6 +834,20 @@ public Order discontinueOrder(Order orderToDiscontinue, String reasonNonCoded, D
...
 	public OrderGroup saveOrderGroup(OrderGroup orderGroup) throws APIException;
 
+	/**
+	 * Saves an order group with a specific order context
+	 *
+	 * @param orderGroup the order group to be saved
+	 * @param orderContext the order context data transfer object containing care setting and 
+	 * the order type to save with the order group
+	 * @return the order group that was saved with the specified order context data
+	 * @since 2.5.0
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/api/OrderService.java:844]: This `@since` annotation is incorrect.  Should be 2.7.0 (as of now).

File: api/src/main/java/org/openmrs/api/OrderService.java [hunk 5]
```diff
@@ -834,6 +834,20 @@ public Order discontinueOrder(Order orderToDiscontinue, String reasonNonCoded, D
...
+	 * @should throw an exception if no order context is provided
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/api/OrderService.java:845]: This `@should` annotation is not correct.  We allow order context to be null.  This should be removed.

File: api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java [hunk 6]
```diff
@@ -167,7 +167,7 @@ public Encounter saveEncounter(Encounter encounter) throws APIException {
 
 		// save the new orderGroups
 		for (OrderGroup orderGroup : encounter.getOrderGroups()) {
-			Context.getOrderService().saveOrderGroup(orderGroup);
+			Context.getOrderService().saveOrderGroup(orderGroup,null);
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java:170]: Why is this change necessary?  Can't we continue to call the existing method?

File: api/src/test/java/org/openmrs/api/OrderServiceTest.java [hunk 7]
```diff
@@ -3963,6 +3963,43 @@ public void saveOrder_shouldAllowARetrospectiveOrderToCloseAnOrderThatExpiredInT
...
+		OrderType orderType = orderService.getOrderType(17);
+		CareSetting careSetting = orderService.getCareSetting(1);
+
+		OrderContext orderContext = new OrderContext();
+		orderContext.setCareSetting(careSetting);
+		orderContext.setOrderType(orderType);
+		OrderGroup result = orderService.saveOrderGroup(orderGroup, orderContext);
+
+		assertEquals(2, result.getOrders().size());
+		assertEquals(17, result.getOrders().get(0).getOrderType().getOrderTypeId());
```
Comment (mseaton) [inline api/src/test/java/org/openmrs/api/OrderServiceTest.java:4000]: Better would be to assert that the orderType matches `orderType` from above.  Then you aren't hard-coding id 17 in 2 places.

File: api/src/test/java/org/openmrs/api/OrderServiceTest.java [hunk 8]
```diff
@@ -3963,6 +3963,43 @@ public void saveOrder_shouldAllowARetrospectiveOrderToCloseAnOrderThatExpiredInT
...
+		assertEquals("OUTPATIENT", result.getOrders().get(1).getCareSetting().getName());
```
Comment (mseaton) [inline api/src/test/java/org/openmrs/api/OrderServiceTest.java:4001]: Better would be to assert that the careSetting matches the `careSetting` from above.  This adds unnecessary complexity.

**Review by mseaton — CHANGES_REQUESTED** [review]
Thanks @IamMujuziMoses - looking good.  A few additional comments, but this seems very close.

File: api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java (hunk 6 above)
Comment (IamMujuziMoses) [reply api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java:170]: yes, this change is not necessary, let me revert

**Review by mseaton — APPROVED** [review]
Looks good to me.  Thanks @IamMujuziMoses
